Implement path finding

Lumbi opened the pull request:
Implemented path finding through the `ActionGraph` using Dijkstra and cleaned up some of the code.

Lumbi committed:
Add include guards to Assert.h

Lumbi committed:
Add ID to Action calculated as hash of name

Lumbi committed:
Remove const from members in Action

To allow implicit copy-constructor and copy-assign operator

Lumbi committed:
Add is_reachable_from(Conditions) API to Action

Lumbi committed:
Remove ActionPath and just use std::vector

Lumbi committed:
Implement path finding and update tests

Lumbi committed:
Remove unnecessary AcionGraph::Node destructor

Lumbi committed:
Mark ActionGraph::Node constructor as explicit

Lumbi committed:
Remove forward declaration

Already forward declared in method signature

Lumbi committed:
Delete ActionGraph copy-assign operator

Because copy-construct is also deleted. Will need to revisit that later.

Lumbi committed:
Use implicit copy-construct for Condition

No memory management so destructor is also not needed

Lumbi committed:
Remove constness of containee in `Conditions` alias

Lumbi committed:
Forward declare type in method signature instead

Lumbi committed:
Mark Goal constructor as explicit

tristanlabelle commented on src/ActionGraph.cpp:
```diff
@@ -55,10 +53,34 @@ const std::vector<ActionGraph::Node*>& ActionGraph::get_nodes() const
+            if (shortest_path.empty()) {
+                shortest_path = path;
+            } else if (path.size() < shortest_path.size()) {
+                shortest_path = path;
```
You could `std::move(path)` and possibly avoid a copy

tristanlabelle commented on src/ActionGraph.cpp:
```diff
@@ -55,10 +53,34 @@ const std::vector<ActionGraph::Node*>& ActionGraph::get_nodes() const
+    // Map nodes to actions
+    std::vector<Action> actions;
+    for (auto&& node : shortest_path) {
+        actions.insert(actions.begin(), node->action);
```
If you want to impress girls, you can do something like:
```
std::transform(rbegin(actions), rend(actions), back_inserter(actions), [](const Node* n) { n->action });
                              OutputIt d_first, UnaryOperation unary_op );

Lumbi replied:
Ah, much better, but I miss `map` 😭

tristanlabelle commented on src/Goal.h:
```diff
@@ -3,14 +3,12 @@
+        explicit Goal(const Conditions&);
 
-        bool is_reachable_by(const Action&) const;
+        bool is_reachable_by(const class Action&) const;
```
I'm not used to seeing a forward declaration in a parameter but that's kind of cool.

tristanlabelle commented on test/Assert.h:
```diff
@@ -2,6 +2,9 @@
+#ifndef ASSERT_H
+#define ASSERT_H
+
 inline void ASSERT(std::function<bool (void)> predicate, const std::string& message)
```
`std::function` will imply an allocation. You would typically write this function as templatized so it can take a straight functor.

tristanlabelle commented on src/ActionGraph.h:
```diff
@@ -6,18 +6,13 @@
-                ~Node();
+                explicit Node(const Action& action);
 
                 const std::vector<const Node*> get_successors() const;
```
The first const is funny since you're returning a copy, not a reference.

Lumbi replied:
derp

tristanlabelle commented on src/PathFinding.h:
```diff
@@ -0,0 +1,91 @@
+)
+{
+    std::vector<Node*> to_visit;
+    std::map<Key, bool> visited;
```
`unordered_map` is usually the preferred default

tristanlabelle replied:
`unordered_set` would be better in this case

tristanlabelle commented on src/PathFinding.h:
```diff
@@ -0,0 +1,91 @@
+
+        // Explore neighbors and calculate distances
+        for (auto&& neighbor : get_neighbors(*nearest)) {
+            if (visited[get_key(*neighbor)] == false) {
```
I think this creates the entry if it doesn't exist so it would make more sense to call `find`

Lumbi replied:
I didn't mind because it defaults to `false` but yeah probably nicer to avoid implicit side-effects

tristanlabelle commented on src/PathFinding.h:
```diff
@@ -0,0 +1,91 @@
+            : std::numeric_limits<float>::infinity();
+    };
+
+    while (to_visit.empty() == false) {
```
` == false` is funny. Why not `!` ?

Lumbi replied:
😂  probably a habit from the JS world and also `!` is so slim and easy to miss

tristanlabelle commented on src/PathFinding.h:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        // Skip if already visited
+        if (visited[get_key(*nearest)]) {
```
Here too, prefer `find` to avoid mutating the object for a lookup.

tristanlabelle commented on src/Action.cpp:
```diff
@@ -13,6 +13,7 @@ Action::Action(const Conditions& preconditions, const Conditions& postconditions
 
 Action::Action(const Name& name, const Conditions& preconditions, const Conditions& postconditions)
     : _name { name },
+      _id { std::hash<std::string>{}(name) },
```
I doubt this strongly guarantees key uniqueness

Lumbi replied:
I join you in your doubt. I'm probably better off just [implementing it myself](http://www.cse.yorku.ca/~oz/hash.html).

tristanlabelle commented on src/PathFinding.h:
```diff
@@ -0,0 +1,91 @@
+std::vector<Node*> find_shortest_path(
+    Node& start,
+    std::function<Key (Node&)> get_key,
+    std::function<std::vector<Node*> (Node&)> get_neighbors,
```
C++ trained me to be alarmed when seeing so many `std::vectors` being created, although you would probably do this in other programming languages.

Lumbi replied:
To be honest, I felt pretty weird spamming `std:vector` but didn't know how to do this more idiomatically

tristanlabelle commented on src/PathFinding.h:
```diff
@@ -0,0 +1,91 @@
+template<typename Node, typename Key>
+std::vector<Node*> find_shortest_path(
+    Node& start,
+    std::function<Key (Node&)> get_key,
```
I'm not used to seeing this many `std::function` used as delegates, but it's possible it's due to the kind of C++ code I usually look at.

If you want to play with C++ 20 concepts, you could have a `PathfindingQuery` concept for a struct with those three methods and typedefs/usings for the node and key types. That'd be cool.

Lumbi committed:
Use replace std::map with std::unordered_map and std::unordered_set

Lumbi committed:
Return ActionGraph::Node successors by reference

Lumbi committed:
Replace for-loop-insert with std::transform

Lumbi committed:
Test path finding with multiple viable paths where last path is the shortes

Lumbi committed:
Fix bug in path finding

Although, functionality-wise I think it didn't matter, it was just not efficient ;)

Lumbi committed:
Use std::move to avoid copy-assignment

Lumbi committed:
Replace std::function with template parameter with type constraint

Lumbi committed:
Replace path finding std::function delegates with concept
